### PR TITLE
A look at C# coding conventions

### LINKED ISSUE #4147: [Languages] A look at C-Sharp coding conventions.
#### Proposed title of article
[Languages] A look at C-Sharp coding conventions.

#### Proposed article introduction
Coding conventions are a set of recommendations for programming style, practices, and methods for each aspect of a program written in a specific programming language. 

Coding conventions serves various purposes as listed below.
1. They make it easier to copy, change, and maintain code.
2. They help readers grasp the code more quickly by allowing them to make assumptions based on prior knowledge.
3. They give the code a uniform appearance so that readers can concentrate on the content rather than the layout.

The standards are what makes a code look neat because it has met almost everything and the probability of it having an error is very low. Such standards include the layout conventions, naming conventions and many others.

#### Key takeaways 
1. Layout conventions.
2. Naming conventions
3. Commenting conventions.

#### Article quality
In this article, I want to provide content that clearly shows that 
there are rules that are applied when programming. Many people just code but I  will have to deeply explain what this rules entails and how how to go about it
#### References
N/A

### PR DESCRIPTION
**Pull request by amosnjoro**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #issue-number`
This closes #4147

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Kayere** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content. Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[A look at C# coding conventions #4659 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7453883/A.look.at.C.coding.conventions.4659.Copyleaks.Report.pdf)

**Comment by amosnjoro** [comment]
@Kayere I am working on it

**Comment by amosnjoro** [comment]
@Kayere  When I am done you will open it so that I can resubmit

**Comment by Kayere** [comment]
No problem.

**Comment by amosnjoro** [comment]
@Kayere  You can reopen so that I can fix it now

**Comment by Dawe-7** [comment]
@amosnjoro, Go ahead.

@Kayere, opening this PR.

**Comment by amosnjoro** [comment]
@Dawe-7 I have updated it,, you can have a look

**Comment by Dawe-7** [comment]
@amosnjoro, before we begin the review process kindly go through your article. There are several mistakes  that i saw as i glanced through your article. Check your article for readability and ensure your article is fully polished and ready for publishing. Make sure to write your article in third person view and not the second person view. 

A few things to add:
1. Introduction and better conclusion - Kindly add an introduction as to what are coding conventions before diving right into the importance as you have done. Improve on your conclusion by summarizing what the article is about and give the reader a few pointers. 
2. Use headings and subheadings - I noticed that you used the `**` asterisks instead of headings and subheadings. Check on this.

Ping me when your article is fully polished.

**Comment by Dawe-7** [comment]
> @Dawe-7 I have done the changes

Review in progress

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+2. They help readers understand the code faster by allowing them to make assumptions based on prior knowledge.
+3. They give the code a unified design so readers may concentrate on the content rather than the layout.
+4. They tell the best way to use the C# programming language in the most ideal manner.
+
+Several coding conventions are used in the C# programming language:
+1. Naming conventions
+2. Layout conventions
+3. Commenting conventions
+
+There are a few naming conventions that should be followed while writing C# code. All information about public elements applies as well to protected and protected internal elements in the following scenarios, which are all supposed to be visible to external callers.
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:32]: Explain the second sentence here

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+```
+voidMain
+```
+
+Another implementation of the camel case is when naming private or internal fields.
+
+```
+camelCasing
+```
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:111]: Is this supposed to be here?

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+```c#
+public T SomeMethod<T>(int someNumber, bool isValid)
+{
+}
+```
+
+Apart from the above conventions, there are other several conventions that a programmer needs to consider and apply while in their daily routine of programming.
+1. The choice of word - Use readable identifier names. Underscores, hyphens, and other non-alphanumeric characters should not be used.
+2. Hungarian notation is neither acceptable nor encouraged.
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:142]: Explain further

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+3. The use of Abbreviations and Acronyms - Do not use acronyms that aren't generally recognized, and even if they are, only use them when essential.
+4. Abbreviations and contractions should not be used as identifier names.
+5. Use `GetWindow` instead of `GetWin`.
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:145]: Explain further here as well

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+### Layout Conventions
+For a clear understanding of the code and its structure, formatting is exercised as a coding convention.
+
+Some of the layout coding conventions are:
+- Use the code editor's default settings that use smart indenting, four-character indents, and tabs that are saved as spaces.
+- Each line should accommodate one statement.
+- Each line should only accommodate one declaration.
+- If the continuation lines are not indented consequently, they are supposed to be indented at a stop of one tab.
+- At least a single blank line is supposed to be added between the property definitions and the method.
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:155]: Explain why this is supposed to be done

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+4. They tell the best way to use the C# programming language in the most ideal manner.
+
+Several coding conventions are used in the C# programming language:
+1. Naming conventions
+2. Layout conventions
+3. Commenting conventions
+
+There are a few naming conventions that should be followed while writing C# code. All information about public elements applies as well to protected and protected internal elements in the following scenarios, which are all supposed to be visible to external callers.
+
+### Pascal's case
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:34]: If pascal and the camel case are naming conventions, kindly put a heading for `Naming conventions` and the the cases should be subheadings of this.

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    string Country,
+    string County,
+    string Village,
+    );
+public struct ValueCoordinate1
+{
+}
+```
+
+Besides to prefixing the name with an I, when naming an interface, use the pascal casing. Customers can tell it's a user interface because of this.
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:57]: Explain further here

**Review by Dawe-7 — CHANGES_REQUESTED** [review]
Nice article @amosnjoro. However, i feel the article is on a light substance. Kindly make the suggestions and try to have better explanations and make the article flow.

**Comment by Dawe-7** [comment]
@amosnjoro, any updates on this PR?

**Comment by hectorkambow** [comment]
@amosnjoro 

 it seems this PR is still quite a bit aways from being completed. 
 
As a rule of thumb - lets please only submit articles (pull requests) with a complete and fully polished article - ready to be published. 
 
All revisions and edits should be completed on your own forked repo (as to not take up room in the queue). 
 
We will be closing this and please re-submit when it is ready to be published as a new PR (and you can add the duplicate label) and ping the reviewer and myself. 

Closing a PR does NOT mean the article can not be published (or fixed) it simply means that the edits, revisions, and fixes will happen outside of the queue.
 
👍 

cc @Dawe-7 let me know if we decide something different.

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 1 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:32]: What I mean in this sentence is that any guidance pertaining elements that are public is also applicable in protected and protected internal elements

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 2 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:111]: I was giving it out as an example but I have realized that I positioned it at the wrong place. I opt to remove it

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 3 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:142]: I have explained in the article above as follows
1. Choice of word- Not every word is used in programming but instead the correct identifiers and so are used together with the proper punctuation eg where and when to put a comma and such
2. Hungarian notation is neither acceptable nor encouraged - This is because a hungarian notation is a coding convention whose variable name depicts what it intends to do and if this is continually used it creates a programming language dependency and makes maintenance more difficult.

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 4 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:145]: I have also gone ahead and elaborated it in the article as follows.
1.  Abbreviations and contractions should not be used as identifier names- This implies that one should not contract any name or word in order to use them as identifiers rather they are supposed to use the identifiers that are clearly identifiable.
2. There was a small typing mistake but I corrected it to look like this `Do not use `GetWin` instead of `GetWindow`This is because it is somehow confusing instead a programmer is advised to use well-known abbreviations eg For UserInterface, one may use UI which is more identifiable.

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 5 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:155]: All this is done not that it is too important but from what the experience I have gained and advice from my seniors, it is done to make sure that your code is clean and neat so that it is readable, a reader will have an easy time following your code and to know what each line entails.

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 6 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:34]: I have resolved that.

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 7 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:57]: What I mean is that beside using pascal casing when naming an interface, one should also prefix the interface name with an I so that it can be easily recognized as an interface

**Comment by amosnjoro** [comment]
@Dawe-7  I have completed making the changes you requested ,, you can reopen

**Comment by amosnjoro** [comment]
@Dawe-7 I have completed making the changes that you had requested, you can now reopen

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+1. The choice of word - Use readable identifier names. Underscores, hyphens, and other non-alphanumeric characters should not be used. The choice of word implies that not any word can be used when programming, instead the proper words should be utilized.
+2. Hungarian notation is neither acceptable nor encouraged.
+A hungarian notation is whereby the name of a variable or function denotes its intent or kind, as well as its type in some languages. This therefore creates a programming language dependency and makes maintenance more difficult.
+3. The use of Abbreviations and Acronyms - Do not use acronyms that aren't generally recognized, and even if they are, only use them when essential.
+4. Abbreviations and contractions should not be used as identifier names. This simply means that when one is writing an identifier, words should not be contracted but instead well-known identifiers should be used.
+5. Do not use `GetWin` instead of `GetWindow`. This is because it is somehow confusing instead a programmer is advised to use well-known abbreviations eg For UserInterface, one may use UI which is more identifiable.
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md]: The 5th convention here doesn't make sense. can you rephrase it or find a way to make it more clear. The first sentence especially has not been written well.

File: content/articles/a-look-at-c#-coding-conventions/index.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+### Layout Conventions
+For a clear understanding of the code and its structure, formatting is exercised as a coding convention.
+
+Some of the layout coding conventions are:
+- Use the code editor's default settings that use smart indenting, four-character indents, and tabs that are saved as spaces.
+- Each line should accommodate one statement.
+- Each line should only accommodate one declaration.
```
Comment (Dawe-7) [inline content/articles/a-look-at-c#-coding-conventions/index.md:150]: explain this two further. why should we accommodate one statement and why we each line should accommodate one declaration?

**Review by Dawe-7 — CHANGES_REQUESTED** [review]
@Kindly ensure all changes requested are effected because there are some which were not.

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 8 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md]: It means that getwin may bring about confusion to a reader when reading but instead getwindow in full may not bring  that confusion because its already in full and well understood.

File: content/articles/a-look-at-c#-coding-conventions/index.md (hunk 9 above)
Comment (amosnjoro) [reply content/articles/a-look-at-c#-coding-conventions/index.md:150]: One statement per line means one is not supposed to explain or to comment about two or more different fields on a single line. Each commenting line should consist a single explanation about a certain say variable, identifier or such.

One declaration per line simply means that declaring multiple variables in a single declaration may bring about confusion about the type of variables and their initial values. This will help the reader to have a smooth flow when reading.

**Comment by hectorkambow** [comment]
@amosnjoro 
Good afternoon and thank you for submitting your topic suggestion.
 
Not a problem - it seems this PR is still quite a bit aways from being completed. 
 
As a rule of thumb - please only submit articles (pull requests) with a complete and fully polished article - ready to be published. 
 
All revisions and edits should be completed on your own forked repo (as to not take up room in the queue). 
 
We will be closing this and please re-submit when it is ready to be published as a **new PR** (and you can add the duplicate label) and ping the reviewer and myself. 

Closing a PR does NOT mean the article can not be published (or fixed) it simply means that the edits, revisions, and fixes will happen outside of the queue.
 
👍

**Comment by amosnjoro** [comment]
@hectorkambow  I had made the required changes and now I was waiting the response from @Dawe-7 . The changes  that he had requested I had completed them.

**Comment by Dawe-7** [comment]
> @hectorkambow I had made the required changes and now I was waiting the response from @Dawe-7 . The changes that he had requested I had completed them.

@amosnjoro, you only left comments giving the explanations requested rather than doing it in the article and pushing them. Kindly take a look at the last comment i made on this PR.

Closing a PR does not mean your article will not be published rather you work on the changes on your own forked repo and create a new PR once done. Kindly make all the requested changes and give better explanations in the article and ping me when you've created a new PR that is fully polished and with the changes effected. 

Thanks.

**Comment by amosnjoro** [comment]
@Dawe-7 I have corrected that

**Comment by Dawe-7** [comment]
> @Dawe-7 I have corrected that

@kindly go ahead and create a new PR with everything set.
